language-plutus-core/src/Language/PlutusCore/Lexer/Type.hs

```diff
@@ -141,3 +144,32 @@ newIdentifier str = do
             let nextU' = Unique $ unUnique nextU + 1
             put (M.insert str nextU ss, nextU')
             pure nextU
+
+
+toStaticBuiltinName :: T.Text -> Maybe StaticBuiltinName
+toStaticBuiltinName = \case
```

Comment:
@effectfully But somewhere we have to convert the syntactic names of the builtins into the relevant constructors (which are what are stored in `allBuiltinNames`).  I had a look, but I can't see that happening anywhere else.  I don't quite see what you mean here.

Reply:
You can do

```haskell
import qualified Data.Map.Strict as Map

allBuiltinNamesMap :: Map.Map String BuiltinName
allBuiltinNamesMap = Map.fromList $ map (\x -> (show $ pretty x, x)) allBuiltinNames
```

and look up syntactic names in `allBuiltinNamesMap`, right?

Reply:
You can do

```haskell
import qualified Data.Map.Strict as Map

allBuiltinNamesMap :: Map.Map String BuiltinName
allBuiltinNamesMap = Map.fromList $ map (\x -> (show $ pretty x, x)) allBuiltinNames
```

and look up syntactic names in `allBuiltinNamesMap`, right?

Reply:
That did cross my mind, but it almost seems the wrong way round:  it'd be quite hard to work out what the concrete syntax for the builtins is without digging quite deeply into the code.  

Having said that, I suppose we'll _have_ to do something like this for dynamic builtins, which will be important if we get to a point where that's all we have.  The parser should somehow be provided with a list of the builtins that are going to be available at execution time so that it can complain if the code contains something unexpected, so that'll have to be derived from a `DynamicBuiltinMeanings` structure or something similar.

Reply:
> That did cross my mind, but it almost seems the wrong way round: it'd be quite hard to work out what the concrete syntax for the builtins is without digging quite deeply into the code.

The concrete syntax for builtins is the one that they get printed as. That's the point of not introducing two different syntaxes which happen to match.

> Having said that, I suppose we'll have to do something like this for dynamic builtins, which will be important if we get to a point where that's all we have. The parser should somehow be provided with a list of the builtins that are going to be available at execution time so that it can complain if the code contains something unexpected, so that'll have to be derived from a DynamicBuiltinMeanings structure or something similar.

Yep.

Reply:
> The concrete syntax for builtins is the one that they get printed as.
I'm not entirely convinced by that: I could equally well say that the concrete syntax for builtins should be the one they get read as!  I think what disturbs me a bit here is that the concrete name for a builtin is rather loosely coupled to the builtin itself via the `Pretty` typeclass.  If you're looking at the code trying to see what the textual form of a builtin is then you have to locate the `Pretty` instances, which are buried away in a file that's far from easy to find, and if you're trying to do it programatically you have to do it indirectly, by calling `show . pretty`  I think I'd be happier if the definitive string representation of a builtin was included directly in the object defining the builtin itself (maybe the `DynamicBuiltinNameDefinition` for example), where it could be accessed easily by both the prettyprinter and the parser (and also by people looking at the code).  Maybe we could think about doing that if and when we get rid of static builtins.

Anyway, never mind: this is perhaps a bit of a theological argument.  I'll get rid of that table as you suggested.

Reply:
> I'm not entirely convinced by that: I could equally well say that the concrete syntax for builtins should be the one they get read as!

That would also be fine. Anything is fine as long as there's a single source of truth.

>  If you're looking at the code trying to see what the textual form of a builtin is then you have to locate the `Pretty` instances, which are buried away in a file that's far from easy to find, and if you're trying to do it programatically you have to do it indirectly, by calling `show . pretty`

I think it's fairly straightforward to get what's happening by looking at the definition of `toStaticBuiltinName` defined in terms of `allBuiltinNamesMap`.

> Maybe we could think about doing that if and when we get rid of static builtins.

If we keep dynamic builtins as mere `Text`, then we won't need to go through `Pretty`. If we introduce an additional type variable, then going through `Pretty` still seems sensible and convenient to me.